felix/wireguard/bootstrap.go

```diff
@@ -17,110 +17,386 @@ package wireguard
 	"k8s.io/apimachinery/pkg/util/wait"
 
 	"github.com/projectcalico/calico/felix/config"
 	"github.com/projectcalico/calico/felix/netlinkshim"
+	apiv3 "github.com/projectcalico/calico/libcalico-go/lib/apis/v3"
 	"github.com/projectcalico/calico/libcalico-go/lib/clientv3"
 	cerrors "github.com/projectcalico/calico/libcalico-go/lib/errors"
 	"github.com/projectcalico/calico/libcalico-go/lib/options"
+	"github.com/projectcalico/calico/libcalico-go/lib/set"
+	"github.com/projectcalico/calico/typha/pkg/discovery"
 )
 
-// BootstrapHostConnectivity forces WireGuard peers with hostencryption enabled to communicate with this node unencrypted.
+const (
+	bootstrapBackoffDuration  = 200 * time.Millisecond
+	bootstrapBackoffExpFactor = 2
+	bootstrapBackoffMax       = 2 * time.Second
+	bootstrapJitter           = 0.2
+	bootstrapMaxRetries       = 5
+)
+
+// BootstrapHostConnectivity forces WireGuard peers with host encryption to communicate with this node unencrypted.
 // This ensures connectivity in scenarios where we have lost our WireGuard config, but will be sent WireGuard traffic
 // e.g. after a node restart, during felix startup, when we need to fetch config from Typha (calico/issues/5125)
-func BootstrapHostConnectivity(configParams *config.Config, getWireguardHandle func() (netlinkshim.Wireguard, error), calicoClient clientv3.Interface) error {
+//
+// Returns the set of programmed peers to validate. Returns nil if wireguard configuration is not present, has just been
+// removed, or host encryption is not enabled.
+func BootstrapHostConnectivity(
+	configParams *config.Config,
+	getNetlinkHandle func() (netlinkshim.Interface, error),
+	getWireguardHandle func() (netlinkshim.Wireguard, error),
+	calicoClient clientv3.Interface,
+) (set.Set, error) {
 	wgDeviceName := configParams.WireguardInterfaceName
 	nodeName := configParams.FelixHostname
 
-	_, dbgBootstrap := os.LookupEnv("FELIX_DBG_WGBOOTSTRAP")
+	logCxt := log.WithFields(log.Fields{
+		"iface":    wgDeviceName,
+		"nodeName": nodeName,
+	})
 
-	if !configParams.WireguardHostEncryptionEnabled && !dbgBootstrap {
-		return nil
+	if !configParams.WireguardHostEncryptionEnabled {
+		// HostEncryption is currently enabled in environments by operator rather than through FelixConfiguration.
+		// This should not change for a given deployment. We only need to handle bootstrapping for clusters where
+		// host encryption is enabled.
+		logCxt.Debug("Host encryption is not enabled - no wireguard bootstrapping required")
+		return nil, nil
 	}
 
-	logCtx := log.WithFields(log.Fields{
-		"iface":    wgDeviceName,
-		"hostName": nodeName,
-		"ref":      "wgBootstrap",
+	logCxt.Debug("Bootstrapping wireguard")
+
+	if !configParams.WireguardEnabled || configParams.WireguardInterfaceName == "" {
+		logCxt.Info("Wireguard is not enabled - ensure no wireguard config")
+		return nil, RemoveWireguardForHostEncryptionBootstrapping(configParams, getNetlinkHandle, calicoClient)
```

Comment:
Why not do this even if `WireguardHostEncryptionEnabled` is `false`?  Seems best to nuke wireguard in that case too.

Reply:
Yeah seems reasonable - was originally trying to reduce blast radius, but this is a perfectly reasonable thing to do.... it should speed up pod connectivity when turning wireguard off.